Enhancement: Add distributed map write transaction support

edwardsmatt opened the pull request:
**Background**
We have a specific requirement precluding us from being able to use Sticky Sessions (we are currently testing with round-robin load balancers); however, we must also maintain session consistency across the all the nodes. 

Under certain load scenarios, we find that the session manager is sometimes unable to find certain session ids in the session map, which is causing some issues with our implementation of the hazelcast tomcat session replication. However, we have observed these issues are resolved if we use Hazelcast Transactions when writing to the session map.

**Enhancement**
Add support to the `HazelcastSessionManager` to enable configuration to use hazelcast transaction support for updates to the session map. The default functionality must not change and continue to operate as documented.

edwardsmatt commented:
Let me know if this is something you'd be interested in investigating and I can add more extensive tests for different configurations. I've currently only added tests for non-sticky session updates at this stage 👍

emre-aydin commented:
Thanks for your contribution @edwardsmatt. This is definitely something we would like to support. 

I will have a look into your PR in a couple of days as we're currently busy with an upcoming release for another module. In the meantime, would you please sign our Contributor Agreement?
https://hazelcast.atlassian.net/wiki/display/COM/Hazelcast+Contributor+Agreement

emre-aydin commented:
verify

devOpsHazelcast commented:
Test FAILed.

emre-aydin commented on tomcat-core/src/main/java/com/hazelcast/session/txsupport/TwoPhaseCommitMapWriteStrategy.java:
```diff
@@ -0,0 +1,86 @@
+    public TwoPhaseCommitMapWriteStrategy(HazelcastInstance hazelcastInstance, String sessionMapName) {
+        this.hazelcastInstance = hazelcastInstance;
+        this.sessionMapName = sessionMapName;
+        transactionOptions = TransactionOptions.getDefault();
```
Since the class is explicitly meant for two-phase commit and it has `TwoPhase` in its name, it might be better to explicitly configure `TransactionOptions.TransactionType#TWO_PHASE` for `transactionOptions`.

edwardsmatt replied:
Good point 👍

emre-aydin commented on tomcat-core/src/main/java/com/hazelcast/session/AbstractHazelcastSessionManager.java:
```diff
@@ -0,0 +1,86 @@
+        return mapWriteStrategy;
+    }
+
+    public void setMapQueryStrategy(MapQueryStrategy mapQueryStrategy) {
```
This method should be private to avoid confusion.

emre-aydin commented on tomcat-core/src/main/java/com/hazelcast/session/AbstractHazelcastSessionManager.java:
```diff
@@ -0,0 +1,86 @@
+        this.mapQueryStrategy = mapQueryStrategy;
+    }
+
+    public void setMapWriteStrategy(MapWriteStrategy mapWriteStrategy) {
```
This one also should be private.

emre-aydin commented on tomcat-core/src/main/java/com/hazelcast/session/AbstractHazelcastSessionManager.java:
```diff
@@ -0,0 +1,86 @@
+     * @param mapName      the name of the hazelcast tomcat session map.
+     * @param readStrategy the readStrategy setting value.
+     */
+    void configureReadStrategy(String mapName, String readStrategy) {
```
`mapName` is not used, can be removed from the parameter list of `configureReadStrategy`.

edwardsmatt commented:
Thanks for the feedback, I was considering making the `configureReadStrategy` method accept the `mapName` to support future extensions and possible read strategies other than default. 

But as per your feedback and the current state of play, I will remove that parameter 👍 

Also, I will remove the `mapReadStrategy` and `mapWriteStrategy` as those setters as they aren't really required.

Also, I figure I may as well add the implementation for `OnePhaseCommitWriteStrategy`. 

I'll will get the updated PR done shortly. Further, if you think it's worth while, I can attempt to add further tests for the other configuration types (i.e. not just P2P `sticky=false` and `deferredWrite=true`)

emre-aydin commented:
> Also, I figure I may as well add the implementation for OnePhaseCommitWriteStrategy.

Sounds good. Can you also please fix the checkstyle warnings? You can see them [here](https://hazelcast-l337.ci.cloudbees.com/job/TomcatSessionManager-pr-builder/32/). Also, you can try it locally by activating `checkstyle` Maven profile.

> Further, if you think it's worth while, I can attempt to add further tests for the other configuration types (i.e. not just P2P sticky=false and deferredWrite=true)

If that's not too much work for you, it's always welcome to have more tests.

edwardsmatt commented:
No worries, I'll see what I can do 👍

edwardsmatt commented:
Hi @emre-aydin, 
I think I've addressed all the feedback you've provided, including sorting out the checkstyle violations. 

I've also added some more tests, and now the PR looks quite large (I apologise for that); however, it is mostly Javadoc and test code now

Let me know if there is anything else you'd like me to take a look at, or if you have any more concerns 👍

emre-aydin commented:
verify

emre-aydin commented:
Thanks for addressing them so quickly @edwardsmatt, will have a look and let you know.

edwardsmatt commented:
Awesome thanks @emre-aydin, I sent the CLA via email a couple of hours ago.. let me know if it doesn’t arrive and I’ll follow up 👍

devOpsHazelcast commented:
Test PASSed.

edwardsmatt commented:
This is just a general question (and not intended to hurry anything along), but if this is approved, how long before we could expect to see it released and published to maven central?

emre-aydin commented:
We can hope to release it some time next week.

edwardsmatt commented:
Hi @dbrimley, I'll check the other issue and provide some more details. 

Essentially we could reproduce it using P2P and two nodes behind a round-robin load balancer (the two instances were running in EC2). We have an older JSP application using the default Tomcat Form Authenticator (although our legacy application has a newer React front end, which is quite chatty with lots of AJAX calls performing session modifications).

I'll see if I can put together a minimum example to demo what we were experiencing and link it on the new issue.

pveentjer commented:
I agree with @dbrimley ; we should investigate why the problem happens without transactions. It seems there is a problem and that should be resolved. And then there would not be any need to add transactions and we do not make the session manager any more complex.

gokhanoner commented:
@edwardsmatt Please check https://github.com/hazelcast/hazelcast-tomcat-sessionmanager/issues/42

This could explain why you this issue under certain load. So instead of introducing transactions, maybe we need to focus on solving the original issue.

devOpsHazelcast commented:
Can one of the admins verify this patch?

Holmistr commented:
verify

alparslanavci commented:
@edwardsmatt, do you still need this feature? We would like to close this PR since the issue is resolved by #61 and this feature will add more complexity to the module. Please let us know.

edwardsmatt commented:
Sorry I missed your message @alparslanavci .... 

We'll check out the features added in #61 to make sure it works for our case and report back 👍
